lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +802,149 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
   return funcOp;
 }
 
+static APInt mulMod(const APInt &_x, const APInt &_y, const APInt &_cmod) {
+  assert(_x.getBitWidth() == _y.getBitWidth() &&
+         "expected same bitwidth of operands");
+  auto intermediateBitwidth = _cmod.getBitWidth() * 2;
+  APInt x = _x.zext(intermediateBitwidth);
+  APInt y = _y.zext(intermediateBitwidth);
+  APInt cmod = _cmod.zext(intermediateBitwidth);
+  APInt res = (x * y).urem(cmod);
+  return res.trunc(_x.getBitWidth());
+}
+
+static DenseElementsAttr constructRoots(const ShapedType &type,
+                                        const APInt &root, unsigned degree,
+                                        const APInt &cmod) {
+  APInt x = root;
+  APInt primRoot = mulMod(root, root, cmod);
+  SmallVector<APInt> vals(degree / 2);
+  for (unsigned i = 0; i < degree / 2; i++) {
+    vals[i] = x;
+    x = mulMod(x, primRoot, cmod);
+  }
+  return DenseElementsAttr::get(type, vals);
+}
+
+static Value fastNTT(ImplicitLocOpBuilder &b, RankedTensorType &inputType,
+                     Value inputVec, APInt root, APInt cmod) {
+  unsigned degree = inputType.getShape()[0];
+  // Basecase of the algorithm
+  if (degree == 1) {
+    return inputVec;
+  }
+
+  // Type of next iteration
+  auto iterType = inputType.clone({degree / 2});
+
+  // Split the input vector into their odd/even components
+  SmallVector<OpFoldResult> size{b.getIndexAttr(degree / 2)};
+  SmallVector<OpFoldResult> stride{b.getIndexAttr(2)};
+  SmallVector<OpFoldResult> offset{b.getIndexAttr(0)};
+
+  auto evens = b.create<tensor::ExtractSliceOp>(iterType, inputVec, offset,
+                                                size, stride);
+  offset = {b.getIndexAttr(1)};
+  auto odds = b.create<tensor::ExtractSliceOp>(iterType, inputVec, offset, size,
+                                               stride);
+
+  // Recursively iterate with root = root^2 and degree / 2
+  APInt rootSquared = mulMod(root, root, cmod);
+  Value evenNTT = fastNTT(b, iterType, evens.getResult(), rootSquared, cmod);
+  Value oddNTT = fastNTT(b, iterType, odds.getResult(), rootSquared, cmod);
+
+  // Construct a dense cmod to allow for elementwise remainder
+  Value cModVec = b.create<arith::ConstantOp>(
+      iterType,
+      DenseElementsAttr::get(
+          iterType, cmod.sextOrTrunc(iterType.getElementTypeBitWidth())));
+
+  // Implementing Equation 23 from https://ieeexplore.ieee.org/document/10177902
+  //
+  // We have that evenNTT = {A_j}, oddNTT = {B_j} and roots = {root^{2j + 1}}
+  // for all 0 <= j < degree/2
+  auto roots = b.create<arith::ConstantOp>(
+      iterType, constructRoots(iterType, root, degree, cmod));
+
+  // Since roots_j * B_j -> [0, cmod^2) then RemUI will compute the modulus
```

Comment:
I think what you're saying here is that overflow is expected? I worry that this will not behave exactly as you anticipate, because the overflow will implicitly compute a different remainder.

E.g., if cmod = 7 but the underlying storage type is `i3`, then `(6^2 mod 8) mod 7 = 4 mod 7 = 4`, but `6^2 mod 7 = 1`, and so the overflow would break the equivalence.

Feel free to file a TODO for later improvement, I think a carefully crafted test would identify if this is a problem.

Reply:
Not quite what I was going for. It was more trying to convey that since both of the inputs to the multiplication are positive within [0, cmod), then we don't need to do the chain of `arith.RemSI -> arith.AddI -> arith.RemUI` with `cmod` to ensure the modulo is computed correctly.

Currently, there is a conversion before we invoke `fastNTT` that will cast it to an intermediate type that has double the bitwidth. For your example that would be `i3 -> i6`, to ensure that there is no overflow.